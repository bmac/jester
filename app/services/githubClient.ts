

export type Gist = {stargazerCount: number};

export const getGistsForUser = async (username: string) => [] as (Gist[]);
