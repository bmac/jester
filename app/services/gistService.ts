import { getGistsForUser } from "~/clients/githubClient";

const truncateLines = (text: string, lines = 10) => {
  return text.split("\n").slice(0, lines).join("\n");
};

export const topGistForUser = async (username: string) => {
  const gists = await getGistsForUser(username);
  return gists
    .sort((a, b) => {
      return b.stargazerCount - a.stargazerCount;
    })
    .map((gist) => {
      if (gist.files.length) {
        gist.files = [gist.files[0]];
      }
      return gist;
    })
    .map((gist) => {
      if (gist.files[0]) {
        gist.files[0].text = truncateLines(gist.files[0].text);
      }
      return gist;
    });
};
