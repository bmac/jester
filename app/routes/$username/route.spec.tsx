import { describe, expect, it, vi } from "vitest";
import { createRemixStub } from "@remix-run/testing";
import type { Gist } from "~/clients/githubClient";
import * as gistService from "~/services/gistService";
import UserName, { headers as routeHeaders, loader, meta } from "./route";
import { HeadersArgs, MetaArgs } from "@remix-run/node";
import { render, screen } from "@testing-library/react";
import { Location } from "@remix-run/react";

const loaderArgs = {
  params: { username: "octocat" },
  request: new Request("https://www.example.com"),
  context: {},
};

describe("loader", () => {
  it("should return the top gists for the username", async () => {
    const gists = [{ id: "gist" } as Gist];
    vi.spyOn(gistService, "topGistForUser").mockResolvedValue(gists);
    const response = await loader(loaderArgs);

    const data = await response.json();

    expect(data.topGists).toEqual(gists);
  });

  it("should reject with a 404 if the user name is invalid", async () => {
    vi.spyOn(gistService, "topGistForUser").mockRejectedValue(
      new Error("no such user"),
    );

    const response = loader({
      ...loaderArgs,
      params: { username: "no_such_user" },
    });

    expect(response).rejects.toMatchObject({
      status: 404,
    });
  });
});

describe("header", () => {
  it("should cache the response", async () => {
    const headers = new Headers(routeHeaders({} as HeadersArgs));

    expect(headers.get("cache-control")).toEqual(
      "public, max-age=3600, s-maxage=3600",
    );
  });
});

describe("meta", () => {
  it("should render social media meta tags", async () => {
    const metaTags = meta({
        params: { username: 'bmac' },
        data: { topGists: [{
            description: 'a gitignore',
            stargazerCount: 3,
            id: '',
            url: 'https://example.com',
            files: [{
                name: '.gitignore',
                text: '.DS_Store',
            }]
        }] },
        location: { pathname: '/bmac'} as Location,
    } as MetaArgs);

    expect(metaTags).toMatchSnapshot();
  });



  it("should render social media meta tags when there are no gists", async () => {
    const metaTags = meta({
        params: { username: 'bmac' },
        data: { topGists: [] },
        location: { pathname: '/bmac'} as Location,
    } as MetaArgs);

    expect(metaTags).toMatchSnapshot();
  });
});

describe("UserName", () => {
  it("should render the cards from the loader", async () => {
    const RemixStub = createRemixStub([
      {
        path: "/",
        Component: UserName,
        loader() {
          return {
            topGists: [
              {
                id: 1,
                files: [],
                description: "my gist",
              },
            ],
          };
        },
      },
    ]);

    render(<RemixStub />);
    await screen.findByText("my gist");
  });

  it("should display a message when topGists is empty", async () => {
    const RemixStub = createRemixStub([
      {
        path: "/",
        Component: UserName,
        loader() {
          return { topGists: [] };
        },
      },
    ]);

    render(<RemixStub />);
    await screen.findByText("Empty Deck");
  });
});
