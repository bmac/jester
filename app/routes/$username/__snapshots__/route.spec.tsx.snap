// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`meta > should render social media meta tags 1`] = `
[
  {
    "title": "Jester — Top gist stars for bmac",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "name": "title",
  },
  {
    "content": "bmac's most popular gists: .gitignore (3 starts) .DS_Store",
    "name": "description",
  },
  {
    "content": "website",
    "property": "og:type",
  },
  {
    "content": "https://jester.codes/bmac",
    "property": "og:url",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "property": "og:title",
  },
  {
    "content": "bmac's most popular gists: .gitignore (3 starts) .DS_Store",
    "property": "og:description",
  },
  {
    "content": "https://jester.codes/bmac/preview.png",
    "property": "og:image",
  },
  {
    "content": "https://jester.codes/bmac",
    "property": "twitter:url",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "property": "twitter:title",
  },
  {
    "content": "bmac's most popular gists: .gitignore (3 starts) .DS_Store",
    "property": "twitter:description",
  },
  {
    "content": "https://jester.codes/bmac/preview.png",
    "property": "twitter:image",
  },
]
`;

exports[`meta > should render social media meta tags when there are no gists 1`] = `
[
  {
    "title": "Jester — Top gist stars for bmac",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "name": "title",
  },
  {
    "content": "bmac's most popular gists:  (0 starts) ",
    "name": "description",
  },
  {
    "content": "website",
    "property": "og:type",
  },
  {
    "content": "https://jester.codes/bmac",
    "property": "og:url",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "property": "og:title",
  },
  {
    "content": "bmac's most popular gists:  (0 starts) ",
    "property": "og:description",
  },
  {
    "content": "https://jester.codes/bmac/preview.png",
    "property": "og:image",
  },
  {
    "content": "https://jester.codes/bmac",
    "property": "twitter:url",
  },
  {
    "content": "Jester — Top gist stars for bmac",
    "property": "twitter:title",
  },
  {
    "content": "bmac's most popular gists:  (0 starts) ",
    "property": "twitter:description",
  },
  {
    "content": "https://jester.codes/bmac/preview.png",
    "property": "twitter:image",
  },
]
`;
